GeneratorInterface/RivetInterface/src/HiggsTemplateCrossSections.cc

```diff
@@ -0,0 +1,549 @@
+      if (prodVtx == nullptr) return false;
+      // for each ancestor, check if it matches any of the input particles
+      for (auto ancestor:particles(prodVtx, HepMC::ancestors)){ 
+	for ( auto part:ptcls ) 
+	  if ( ancestor==part.genParticle() ) return true;
+      }
+      // if we get here, no ancetor matched any input particle
+      return false; 
+    }
+    
+    /// @brief Whether particle p originates from p2
+    bool originateFrom(const Particle& p, const Particle& p2 ) { 
+      Particles ptcls = {p2}; return originateFrom(p,ptcls);
+    }
+    
+    /// @brief Checks whether the input particle has a child with a given PDGID 
+    bool hasChild(const GenParticle *ptcl, int pdgID) {
+      for (auto child:Particle(*ptcl).children())
+        if (child.pdgId()==pdgID) return true;
+      return false;
+    }
+    
+    /// @brief Checks whether the input particle has a parent with a given PDGID 
+    bool hasParent(const GenParticle *ptcl, int pdgID) {
+      for (auto parent:particles(ptcl->production_vertex(),HepMC::parents))
+        if (parent->pdg_id()==pdgID) return true;
+      return false;
+    }
+
+    /// @brief Return true is particle decays to quarks
+    bool quarkDecay(const Particle &p) {
+      for (auto child:p.children())
+        if (PID::isQuark(child.pdgId())) return true;
+      return false;
+    }
+    
+    /// @brief Returns the classification object with the error code set.
+    ///        Prints an warning message, and keeps track of number of errors
+    HiggsClassification error(HiggsClassification &cat, HTXS::ErrorCode err, 
+			      std::string msg="", int NmaxWarnings=20) {
+      // Set the error, and keep statistics
+      cat.errorCode = err;
+      ++m_errorCount[err];
+
+      // Print warning message to the screen/log
+      static int Nwarnings = 0;
+      if ( msg!="" && ++Nwarnings < NmaxWarnings )
+	  MSG_WARNING(msg);
+
+      return cat;
+    }
+    /// @}
+
+    /// @brief Main classificaion method.
+    HiggsClassification classifyEvent(const Event& event, const HTXS::HiggsProdMode prodMode ) {
+
+      if (m_HiggsProdMode==HTXS::UNKNOWN) m_HiggsProdMode = prodMode; 
+      
+      // the classification object
+      HiggsClassification cat;
+      cat.prodMode = prodMode;
+      cat.errorCode = HTXS::UNDEFINED;
+      cat.stage0_cat = HTXS::Stage0::UNKNOWN;
+      cat.stage1_cat_pTjet25GeV = HTXS::Stage1::UNKNOWN;
+      cat.stage1_cat_pTjet30GeV = HTXS::Stage1::UNKNOWN;
+
+      if (prodMode == HTXS::UNKNOWN) 
+	return error(cat,HTXS::PRODMODE_DEFINED,
+		     "Unkown Higgs production mechanism. Cannot classify event."                                                                       
+		     " Classification for all events will most likely fail.");
+
+      /*****
+       * Step 1. 
+       *  Idenfify the Higgs boson and the hard scatter vertex
+       *  There should be only one of each.
+       */
+
+      GenVertex *HSvtx = event.genEvent()->signal_process_vertex();
```

Comment:
can it be a const &?

Reply:
something like
`const GenVertex &HSvtx = event....`
?

Reply:
you missed a star
=const GenVertex &HSvtx = * event....=

Reply:
tried but leads to this error
http://perrozzi.web.cern.ch/perrozzi/cmssw_pull_20683.log
tried to change a few things but didn't work... suggestions?

Reply:
well if the pointer is checked later for validity then the answer to david question above is "no"... unless you want to check there before assignment if the pointer is null, then assign to a ref (but I do not see the point...)